src/c-writer.cc

```diff
@@ -2489,7 +2552,12 @@ void CWriter::Write(const ExprList& exprs) {
             module_->tables[module_->GetTableIndex(inst->table_index)];
         const ElemSegment* src_segment =
             module_->GetElemSegment(inst->segment_index);
-        Write("table_init(", ExternalInstancePtr(dest_table->name), ", ");
+        if (dest_table->elem_type != src_segment->elem_type) {
+          WABT_UNREACHABLE;
+        }
```

Comment:
Should this be an assert instead?   (I guess it would be validation error if this were true.. so it should be impossible for that to happen here?)

Reply:
I agree this should be a validation error, but my understanding is that the validator doesn't currently catch this one. :-( Per https://github.com/WebAssembly/wabt/issues/1612

Reply:
Still, why not use assert here?

Reply:
My thinking was that an assert isn't strong enough here, because cmake disables asserts on the Release builds. :-/ If we made this an assert, then a user could take an invalid module like this:
```wat
(module
 (table 1 externref)
 (func $f)
 (elem (i32.const 0) func $f)
)
```
and successfully pass it through wat2wasm and wasm2c (release build) and all they would get is a warning when compiling the generated code (`warning: passing argument 1 of ‘funcref_table_init’ from incompatible pointer type; note: expected ‘wasm_rt_funcref_table_t *’ but argument is of type ‘wasm_rt_externref_table_t *’`), and unless they compile everything with `-Werror`, the module would end up with the ability to create a garbage externref that the host never gave it. This feels kind of icky to have as a known bug...

Reply:
True. but UNREACHABLE isn't much better as an end user experience is it?  Wouldn't it manifest as a crash, which then would require debugger to figure out what happens?  IIRC it doesn't even given a line number.

I guess the real fix is to update the validator.    How we add a TODO that says something like "This should really be and assert since the validator should catch this, but currently it doesn't?"

Reply:
Done.